### PR TITLE
[WIP] Web3 react exploration

### PR DESCRIPTION
**Pull request by georgeweiler**

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -14,8 +14,10 @@ import bg from './img/bg.jpg';
 import NoMatch from './pages/noMatch';
 import CrucibleDetail from './pages/crucibleDetail';
 import CrucibleMinting from './pages/crucibleMinting';
+import { useWeb3 } from './context/web3';
 
 const App: FC = () => {
+  useWeb3();
```
Comment (nuurcodes) [inline src/App.tsx:20]: Do we need this here?

File: src/components/minting/mintingGuide.tsx [hunk 2]
```diff
@@ -14,19 +14,17 @@ import { Alert } from '@chakra-ui/alert';
 import { useWeb3 } from '../../context/web3';
 import { config } from '../../config/variables';
 import WelcomeToast from '../shared/welcomeToast';
+import { InjectedConnector } from '@web3-react/injected-connector';
+import { useWeb3React } from '@web3-react/core';
+// @ts-ignore
+const injected = new InjectedConnector({ supportedChainIds: [1, 3, 4, 5, 42] });
```
Comment (nuurcodes) [inline src/components/minting/mintingGuide.tsx:20]: Probably only want to support 1 (mainnet) and 4 (Rinkeby) for now

File: src/context/web3/web3.tsx [hunk 3]
```diff
@@ -1,29 +1,16 @@
-import React, {
```
Comment (nuurcodes) [inline src/context/web3/web3.tsx:1]: I would probably delete this file as there most there is too much unused code
We can write a simpler hook for fetching token balances
We can also use this hook for listening to network changes:
https://github.com/NoahZinsmeister/web3-react/blob/v6/example/hooks.ts

File: src/index.tsx [hunk 4]
```diff
@@ -16,47 +16,51 @@ import { ApolloProvider } from '@apollo/client';
...
-                </ChakraProvider>
-              </LpStatsProvider>
-            </CruciblesProvider>
-          </NetworkStatsProvider>
-        </NotifyProvider>
-      </ApolloProvider>
-    </Web3Provider>
+    <Web3ReactProvider getLibrary={getLibrary}>
+      <Web3Provider
+        tokensToWatch={{
```
Comment (nuurcodes) [inline src/index.tsx:33]: This will have to be dynamic in the near future, maybe we can handle it later

File: src/context/network/stats.tsx [hunk 5]
```diff
@@ -21,22 +21,18 @@ const NetworkStats = createContext<NetworkStatsContextType | undefined>(
...
 
   useEffect(() => {
-    if (
-      provider &&
-      wallet &&
-      address &&
-      network === process.env.REACT_APP_NETWORK_ID
-    ) {
-      const signer = provider?.getSigner();
+    if (library && account && chainId === process.env.REACT_APP_NETWORK_ID) {
```
Comment (nuurcodes) [inline src/context/network/stats.tsx:29]: We won't be pulling the network id from the environment variable anymore, I think we can safely remov `chainId === process.env.REACT_APP_NETWORK_ID`

File: src/contracts/useContracts.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
+import useConfigVariables from '../hooks/useConfigVariables';
```
Comment (nuurcodes) [inline src/contracts/useContracts.ts:1]: Nice 👍

File: src/hooks/useCustomRPCNetworkListener.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+  useEffect(() => {
+    // @ts-ignore
+    const { ethereum } = window;
+    if (ethereum && !active && !error) {
+      // handles RPC reinitialization when user changes to custom network like Taichi
+      // @ts-ignore
+      const handleRPCisReady = (connectInfo) => {
+        if (connectInfo?.chainId === '0x1') {
+          activate(injectedConnector);
```
Comment (nuurcodes) [inline src/hooks/useCustomRPCNetworkListener.ts:17]: Should we activate for Rinkeby too? Not just mainnet

Comment (nuurcodes) [inline src/hooks/useCustomRPCNetworkListener.ts:17]: Should we activate for Rinkeby too? Not just mainnet

File: src/App.tsx [hunk 8]
```diff
@@ -14,60 +14,73 @@ import bg from './img/bg.jpg';
 import NoMatch from './pages/noMatch';
 import CrucibleDetail from './pages/crucibleDetail';
 import CrucibleMinting from './pages/crucibleMinting';
+import { useTokens } from './context/tokens';
 import MobileLayover from './components/modals/MobileLayover';
+import ReduxModal from './components/modals/ReduxModal';
```
Comment (nuurcodes) [inline src/App.tsx:19]: Something like `ModalRoot` could be clearer than `ReduxModal`

File: src/store/hooks.ts [hunk 9]
```diff
@@ -0,0 +1,6 @@
+import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
+import type { RootState, AppDispatch } from './store';
+
+// Use throughout your app instead of plain `useDispatch` and `useSelector`
+export const useAppDispatch = () => useDispatch<AppDispatch>();
```
Comment (nuurcodes) [inline src/store/hooks.ts:5]: Nice 👍

**Review by nuurcodes — APPROVED** [review]
Looks great, left some minor comments

File: src/components/minting/mintingForm.tsx [hunk 10]
```diff
@@ -1,14 +1,14 @@
...
 import MintingHelper from './mintingHelper';
 
 const MintingForm: FC = () => {
-  const { tokens } = useWeb3();
-  const { lpTokenAddress } = config;
+  const { tokens } = useTokens();
+  const { lpTokenAddress } = useConfigVariables();
 
   return (
     <Box>
```
Comment (nuurcodes) [inline src/components/minting/mintingForm.tsx:14]: On the line below we should use:
`tokens[lpTokenAddress]?.balance?.lte(0)`

File: src/hooks/useCustomRPCNetworkListener.ts (hunk 7 above)
Comment (georgeweiler) [reply src/hooks/useCustomRPCNetworkListener.ts:17]: I think we just need to worry about mainnet here to catch taichi. Rinkeby won't disconnect and then try to reconnect the RPC

**Pull request merged**